pkg/controller/cloud/nodecontroller.go

```diff
@@ -91,10 +110,95 @@ func (cnc *CloudNodeController) Run() {
 	go func() {
 		defer utilruntime.HandleCrash()
 
+		instances, ok := cnc.cloud.Instances()
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("failed to get instances from cloud provider"))
+			return
+		}
+
+		// Start a loop to periodically update the node addresses obtained from the cloud
 		go wait.Until(func() {
-			nodes, err := cnc.kubeClient.Core().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
+			nodes, err := cnc.kubeClient.CoreV1().Nodes().List(metav1.ListOptions{ResourceVersion: "0"})
 			if err != nil {
 				glog.Errorf("Error monitoring node status: %v", err)
+				return
+			}
+
+			for i := range nodes.Items {
+				node := &nodes.Items[i]
+
+				// Do not process nodes that are still tainted
+
+				cloudTaint := getCloudTaint(node.Spec.Taints)
+
+				if cloudTaint != nil {
+					glog.V(5).Infof("This node %s is still tainted. Will not process.", node.Name)
+					continue
+				}
+
+				nodeAddresses, err := instances.NodeAddressesByProviderID(node.Spec.ProviderID)
+				if err != nil {
+					nodeAddresses, err = instances.NodeAddresses(types.NodeName(node.Name))
+					if err != nil {
+						glog.Errorf("failed to get node address from cloud provider: %v", err)
+						continue
+					}
+				}
+
+				// Check if a hostname address exists in the cloud provided addresses
+				hostnameExists := false
+				for i := range nodeAddresses {
+					if nodeAddresses[i].Type == v1.NodeHostName {
+						hostnameExists = true
+					}
+
+				}
+				// If hostname was not present in cloud provided addresses, use the hostname
+				// from the existing node (populated by kubelet)
+				if !hostnameExists {
+					for _, addr := range node.Status.Addresses {
+						if addr.Type == v1.NodeHostName {
+							nodeAddresses = append(nodeAddresses, addr)
+						}
+					}
+				}
+				// If nodeIP was suggested by user, ensure that
+				// it can be found in the cloud as well (consistent with the behaviour in kubelet)
+				if providedIP, ok := node.ObjectMeta.Labels[metav1.LabelProvidedIPAddr]; ok {
+					var nodeIP *v1.NodeAddress
+					for i := range nodeAddresses {
+						if nodeAddresses[i].Address == providedIP {
+							nodeIP = &nodeAddresses[i]
+							break
+						}
+					}
+					if nodeIP == nil {
+						glog.Errorf("failed to get node address from cloudprovider that matches ip: %v", nodeIP)
+						continue
+					}
+					nodeAddresses = []v1.NodeAddress{
```

Comment:
Wouldn't this be the same as `[]v1.NodeAddress{*nodeIP}`?
Also, why use a pointer for nodeIP?